service/history/workflow/mutable_state_impl_restart_activity_test.go

```diff
@@ -0,0 +1,370 @@
+	taskGeneratorMock.EXPECT().GenerateActivityRetryTasks(s.activity.ScheduledEventId).Return(e)
+	s.mutableState.taskGenerator = taskGeneratorMock
+
+	ts := TimeSourceStub{}
+	s.mutableState.timeSource = ts
+
+	s.backoffCalculatorStub.onNextCallReturn(time.Second, enumspb.RETRY_STATE_IN_PROGRESS)
+	state, err := s.mutableState.RetryActivity(s.activity, s.activityFailure())
+	s.assert.Error(err, e.Error())
+	s.assert.Equal(
+		enumspb.RETRY_STATE_INTERNAL_SERVER_ERROR,
+		state,
+		"failure to generate task should produce RETRY_STATE_INTERNAL_SERVER_ERROR got %v",
+		state,
+	)
+	s.assertActivityWasNotScheduled(s.activity, "with failing task generator")
+}
+
+func (s *retrySuite) TestRetryActivity_when_workflow_is_not_mutable_should_fail() {
+	s.mutableState.executionState.State = enumsspb.WORKFLOW_EXECUTION_STATE_COMPLETED
+
+	state, err := s.mutableState.RetryActivity(s.activity, s.activityFailure())
+
+	s.assert.Error(ErrWorkflowFinished, err.Error(), "when workflow finished should get error stating it")
+	s.assert.Equal(enumspb.RETRY_STATE_INTERNAL_SERVER_ERROR, state)
+	s.assertActivityWasNotScheduled(s.activity, "for non-mutable workflow")
+	s.assertNoChange(s.activity, "activity should not change if it is not restarted")
+}
+
+const getBackoffIntervalParametersFormat = "now(%v):currentAttempt(%v):maxAttempts(%v):initInterval(%v):maxInterval(%v):expirationTime(%v):backoffCoefficient(%v):failure(%v):nonRetryable(%v)"
+
+type getBackoffIntervalStub struct {
+	expected string
+	recorded string
+	duration time.Duration
+	state    enumspb.RetryState
+}
+
+func (nbis *getBackoffIntervalStub) getBackoffInterval(
+	now time.Time,
+	currentAttempt int32,
+	maxAttempts int32,
+	initInterval *durationpb.Duration,
+	maxInterval *durationpb.Duration,
+	expirationTime *timestamppb.Timestamp,
+	backoffCoefficient float64,
+	failure *failurepb.Failure,
+	nonRetryableTypes []string,
+) (time.Duration, enumspb.RetryState) {
+	nbis.recorded = fmt.Sprintf(
+		getBackoffIntervalParametersFormat,
+		now,
+		currentAttempt,
+		maxAttempts,
+		initInterval,
+		maxInterval,
+		expirationTime,
+		backoffCoefficient,
+		failure,
+		nonRetryableTypes)
+	return nbis.duration, nbis.state
+}
+
+func (nbis *getBackoffIntervalStub) onNextCallExpect(
+	now time.Time,
+	currentAttempt int32,
+	maxAttempts int32,
+	initInterval *durationpb.Duration,
+	maxInterval *durationpb.Duration,
+	expirationTime *timestamppb.Timestamp,
+	backoffCoefficient float64,
+	failure *failurepb.Failure,
+	nonRetryableTypes []string,
+) {
+	nbis.expected = fmt.Sprintf(
+		getBackoffIntervalParametersFormat,
+		now,
+		currentAttempt,
+		maxAttempts,
+		initInterval,
+		maxInterval,
+		expirationTime,
+		backoffCoefficient,
+		failure,
+		nonRetryableTypes,
+	)
+}
+
+func (nbis *getBackoffIntervalStub) onNextCallReturn(duration time.Duration, state enumspb.RetryState) {
+	nbis.duration = duration
+	nbis.state = state
+}
+
+type TimeSourceStub struct {
```

Comment:
FWIW we do have a fake time source already defined in `EventTimeSource`. Not sure if it could be used here, but if so it might make maintaining these tests a little easier down the road.

Reply:
https://github.com/temporalio/temporal/blob/ce3f42e7077b742375ff93f2353e143cf70e5d32/common/clock/event_time_source.go#L33-L36